pkg/controller/certificaterequests/vault/vault.go

```diff
@@ -0,0 +1,133 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vault
+
+import (
+	"context"
+
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/record"
+
+	apiutil "github.com/jetstack/cert-manager/pkg/api/util"
+	"github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha1"
+	controllerpkg "github.com/jetstack/cert-manager/pkg/controller"
+	"github.com/jetstack/cert-manager/pkg/controller/certificaterequests"
+	crutil "github.com/jetstack/cert-manager/pkg/controller/certificaterequests/util"
+	"github.com/jetstack/cert-manager/pkg/internal"
+	vaultinternal "github.com/jetstack/cert-manager/pkg/internal/vault"
+	"github.com/jetstack/cert-manager/pkg/issuer"
+	logf "github.com/jetstack/cert-manager/pkg/logs"
+	"github.com/jetstack/cert-manager/pkg/util/pki"
+)
+
+const (
+	CRControllerName = "certificaterequests-issuer-vault"
+)
+
+type Vault struct {
+	// used to record Events about resources to the API
+	recorder      record.EventRecorder
+	secretsLister corelisters.SecretLister
+	helper        issuer.Helper
+
+	vaultFactory internal.VaultFactory
+}
+
+func init() {
+	// create certificate request controller for vault issuer
+	controllerpkg.Register(CRControllerName, func(ctx *controllerpkg.Context) (controllerpkg.Interface, error) {
+		vault := NewVault(ctx)
+
+		controller := certificaterequests.New(apiutil.IssuerVault, vault)
+
+		c, err := controllerpkg.New(ctx, CRControllerName, controller)
+		if err != nil {
+			return nil, err
+		}
+
+		return c.Run, nil
+	})
+}
+
+func NewVault(ctx *controllerpkg.Context) *Vault {
+	return &Vault{
+		recorder:      ctx.Recorder,
+		secretsLister: ctx.KubeSharedInformerFactory.Core().V1().Secrets().Lister(),
+		helper: issuer.NewHelper(
+			ctx.SharedInformerFactory.Certmanager().V1alpha1().Issuers().Lister(),
+			ctx.SharedInformerFactory.Certmanager().V1alpha1().ClusterIssuers().Lister(),
+		),
+		vaultFactory: vaultinternal.New,
+	}
+}
+
+func (v *Vault) Sign(ctx context.Context, cr *v1alpha1.CertificateRequest, issuerObj v1alpha1.GenericIssuer) (*issuer.IssueResponse, error) {
+	log := logf.FromContext(ctx, "sign")
+	reporter := crutil.NewReporter(cr, v.recorder)
+
+	_, err := pki.DecodeX509CertificateRequestBytes(cr.Spec.CSRPEM)
+	if err != nil {
+		message := "Failed to decode CSR in spec"
+
+		reporter.Failed(err, "ErrorParsingCSR", message)
+		log.Error(err, message)
+
+		return nil, nil
+	}
+
+	client, err := v.vaultFactory(cr.Namespace, v.secretsLister, issuerObj)
+	if err != nil {
+		log = log.WithValues(
+			logf.RelatedResourceNameKey, cr.Spec.IssuerRef.Name,
+			logf.RelatedResourceKindKey, cr.Spec.IssuerRef.Kind,
+		)
+
+		if k8sErrors.IsNotFound(err) {
+			message := "Required secret resource not found"
+
+			reporter.Pending(err, "MissingSecret", message)
+			log.Error(err, message)
+
+			return nil, nil
```

Comment:
The secret that is missing here is named on the *Issuer* resource and not the *CertificateRequest*. This means we *won't* automatically re-sync when/if it does get created. That said, if the secret doesn't exist then the Vault issuer *should* have also noticed this and updated the `conditions` on the Issuer, which means we *will* be resynced, so this might be okay 😄